File: app/src/main/java/com/commit451/adapterlayout/sample/CheeseAdapter.java
package com.commit451.adapterlayout.sample;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import com.commit451.adapterlayout.AdapterLayout;

import java.util.ArrayList;
import java.util.Collection;

/**
 * Adapter for the recyclerview, which holds cheeses
 */
public class CheeseAdapter extends RecyclerView.Adapter<CheeseViewHolder> {

    private Listener listener;
    private ArrayList<Cheese> values;

    public CheeseAdapter(Listener listener) {
        this.listener = listener;
        values = new ArrayList<>();
    }

    public void setData(Collection<Cheese> cheeses) {
        values.clear();
        values.addAll(cheeses);
        notifyDataSetChanged();
    }

    public void add(Cheese cheese) {
        values.add(cheese);
        notifyItemInserted(values.size()-1);
    }

    public void removeLast() {
        if (!values.isEmpty()) {
            int removeIndex = values.size() - 1;
            values.remove(removeIndex);
            notifyItemRemoved(removeIndex);
        }
    }

    public void changeMiddle() {
        if (!values.isEmpty()) {
            int index = values.size()/2;
            values.get(index).name = "Swiss";
            notifyItemChanged(index);
        }
    }

    public void changeAll() {
        if (!values.isEmpty()) {
            for (Cheese cheese : values) {
                cheese.name = "Swiss";
            }
            notifyItemRangeChanged(0, values.size());
        }
    }

    public void clear() {
        if (!values.isEmpty()) {
            int size = values.size();
            values.clear();
            notifyItemRangeRemoved(0, size);
        }
    }

    @Override
    public CheeseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final CheeseViewHolder holder = CheeseViewHolder.inflate(parent);
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //Takes the place of holder.getAdapterPosition()
                int position = AdapterLayout.getAdapterPosition(holder);
                Cheese cheese = getItemAt(position);
                listener.onItemClicked(cheese);
            }
        });
        return holder;
    }

    @Override
    public void onBindViewHolder(final CheeseViewHolder holder, int position) {
        Cheese cheese = getItemAt(position);
        holder.bind(cheese);
    }

    @Override
    public int getItemCount() {
        return values.size();
    }

    private Cheese getItemAt(int position) {
        return values.get(position);
    }

    public interface Listener {
        void onItemClicked(Cheese cheese);
    }
}

File: app/src/main/java/com/commit451/adapterlayout/sample/CheeseAdapter.kt
package com.commit451.adapterlayout.sample

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.commit451.adapterlayout.AdapterLayout

/**
 * Adapter for the recyclerview, which holds cheeses
 */
class CheeseAdapter(private val listener: (cheese: Cheese) -> Unit) : RecyclerView.Adapter<CheeseViewHolder>() {

    private val values = mutableListOf<Cheese>()

    fun setData(cheeses: Collection<Cheese>) {
        values.clear()
        values.addAll(cheeses)
        notifyDataSetChanged()
    }

    fun add(cheese: Cheese) {
        values.add(cheese)
        notifyItemInserted(values.size - 1)
    }

    fun removeLast() {
        if (values.isNotEmpty()) {
            val removeIndex = values.size - 1
            remove(removeIndex)
        }
    }

    fun remove(index: Int) {
        values.removeAt(index)
        notifyItemRemoved(index)
    }

    fun changeMiddle() {
        if (values.isNotEmpty()) {
            val index = values.size / 2
            values[index].name = "Swiss"
            notifyItemChanged(index)
        }
    }

    fun changeAll() {
        if (values.isNotEmpty()) {
            for (cheese in values) {
                cheese.name = "Swiss"
            }
            notifyItemRangeChanged(0, values.size)
        }
    }

    fun clear() {
        if (values.isNotEmpty()) {
            val size = values.size
            values.clear()
            notifyItemRangeRemoved(0, size)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CheeseViewHolder {
        val holder = CheeseViewHolder.inflate(parent)
        holder.itemView.setOnClickListener {
            //Takes the place of holder.getAdapterPosition()
            val position = AdapterLayout.getAdapterPosition(holder)
            val cheese = getItemAt(position)
            listener.invoke(cheese)
        }
        holder.buttonRemove.setOnClickListener {
            val position = AdapterLayout.getAdapterPosition(holder)
            remove(position)
        }
        return holder
    }

    override fun onBindViewHolder(holder: CheeseViewHolder, position: Int) {
        val cheese = getItemAt(position)
        holder.bind(cheese)
    }

    override fun getItemCount(): Int {
        return values.size
    }

    private fun getItemAt(position: Int): Cheese {
        return values[position]
    }
}
